### PR TITLE
Adding community admin page

### PR DESCRIPTION
**Pull request by SarahBA**

### CODE REVIEW

File: webui/src/data/server.js [hunk 1]
```diff
@@ -9,11 +9,17 @@ const urlRoot = ""; // window.location.origin;
...
-    root()                           { return `${urlRoot}/api/` },
+    root()                            { return `${urlRoot}/api/` },
 
     user()                            { return `${urlRoot}/api/user` },
     userTokens()                      { return `${urlRoot}/api/user/tokens` },
 
+    users(queryString)                { return `${urlRoot}/api/users` + (queryString ? `?q=${queryString}` : ``) }, 
+    // userRoles(id)                     { return `${urlRoot}/api/users/${id}/roles` },
+    // rolesList(id)                     { return `${urlRoot}/api/roles` + (id ? `?q=${id}` : ``) },
+    rolesuers(id)                     { return `${urlRoot}/api/roles/${id}/users` },
```
Comment (emanueldima) [inline webui/src/data/server.js]: imprecise name, please rename this to `userListWithRole`

File: webui/src/data/server.js [hunk 2]
```diff
@@ -9,11 +9,17 @@ const urlRoot = ""; // window.location.origin;
...
+    updateRole(roleid, userid)        { return `${urlRoot}/api/roles/${roleid}/users/${userid}` },
```
Comment (emanueldima) [inline webui/src/data/server.js]: imprecise name, please rename this to `userWithRole`

File: webui/src/data/server.js [hunk 3]
```diff
@@ -285,11 +291,30 @@ class ServerCache {
 
             languages: null,
             disciplines: null,
+
+            currentCommunityRoles: {}, // a dictionary of current community's roles name and ID
```
Comment (emanueldima) [inline webui/src/data/server.js]: It's better not to count on anything being 'current', so let's make this a dictionary of communities to dictionary of roles to users, e.g.: 

communityRoles: {
    role: {user, user, user}
}

File: webui/src/data/server.js [hunk 4]
```diff
@@ -285,11 +291,30 @@ class ServerCache {
...
+            communityUsers: {}, // list of users belong to a specific community
         });
 
         this.store.setIn(['communities'], OrderedMap());
 
         this.getters = {};
+		
+        this.getters.currentCommunityRoles = new Pool(communityID => new Getter(
```
Comment (emanueldima) [inline webui/src/data/server.js]: rename to communityRoles

File: webui/src/data/server.js [hunk 5]
```diff
@@ -285,11 +291,30 @@ class ServerCache {
...
+            apiUrls.community(communityID), null,
+            (results) => {
+                this.store.setIn(['currentCommunityRoles'], fromJS(results.roles));
```
Comment (emanueldima) [inline webui/src/data/server.js]: change to `this.store.setIn(['communityRoles', communityID], `

File: webui/src/data/server.js [hunk 6]
```diff
@@ -285,11 +291,30 @@ class ServerCache {
...
+            } ,
+            (xhr) => this.store.setIn(['currentCommunityRoles'], new Error(xhr)),
```
Comment (emanueldima) [inline webui/src/data/server.js]: change to this.store.setIn(['communityRoles', communityID],

File: webui/src/data/server.js [hunk 7]
```diff
@@ -706,14 +731,76 @@ class ServerCache {
...
             params: data,
             successFn: successFn,
             errorFn: errorFn,
         });
     }
+
+
+    // List roles of a specific community
+    getCommunityRoles(id){
+    	if(this.store.getIn(['currentCommunityRoles']).size == 0){
```
Comment (emanueldima) [inline webui/src/data/server.js]: change to `(this.store.getIn(['currentCommunityRoles', id]`

File: webui/src/data/server.js [hunk 8]
```diff
@@ -706,14 +731,76 @@ class ServerCache {
...
         });
     }
+
+
+    // List roles of a specific community
+    getCommunityRoles(id){
+    	if(this.store.getIn(['currentCommunityRoles']).size == 0){ 
+    		this.getters.currentCommunityRoles.get(id).fetch();
+        }
+       	return this.store.getIn(['currentCommunityRoles']);
```
Comment (emanueldima) [inline webui/src/data/server.js]: change to `return this.store.getIn(['currentCommunityRoles', id]);`

File: webui/src/components/user.jsx [hunk 9]
```diff
@@ -92,6 +93,12 @@ export const UserProfile = React.createClass({
             return this.renderNoUser();
         }
         const roles = user.get('roles');
+
+        const communitiesListTemp = serverCache.getCommunities();
+        var communitiesList = communitiesListTemp.reduce(function(map, community){
+                map[community.get('id').replace(new RegExp("-",'g'),"")] = community.get('name');
```
Comment (emanueldima) [inline webui/src/components/user.jsx]: bad indentation

File: webui/src/components/user.jsx [hunk 10]
```diff
@@ -103,9 +110,12 @@ export const UserProfile = React.createClass({
                     </div>
                     <div className="row">
                         <h3>Roles</h3>
-                        {roles && roles.count() ?
-                            roles.map(r => <li key={r.get('name')}>{r.get('description')}</li>)
-                            : <p>You have no assigned roles</p>
+                        {roles && roles.count() && typeof(communitiesList) !== "undefined" ?
+                            roles.map(r => <li key={r.get('name')}>
```
Comment (emanueldima) [inline webui/src/components/user.jsx]: difficult to follow, please extract to separate function

File: webui/src/components/user.jsx [hunk 11]
```diff
@@ -103,9 +110,12 @@ export const UserProfile = React.createClass({
...
+                                                {r.get('description')}  {r.get('name').includes(':admin')? 
+                                                <Link to={`/communities/${communitiesList[r.get('name').replace(new RegExp("(^com:|:[^:]*$)",'g'),"")]}/communityadminpage`}>(admin page)</Link>: ""}
```
Comment (emanueldima) [inline webui/src/components/user.jsx]: difficult to follow, please extract to separate function

File: webui/src/components/community_admin.jsx [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+                    },
+                    () => {
+                        notifications.danger("An error occured while trying to add the new role to the user. ");
+                    }    
+                );
+        }
+    },
+
+    render() {
+        const communityName = this.props.location.pathname.replace(new RegExp("/communities/|/communityadminpage",'g'),"");
```
Comment (emanueldima) [inline webui/src/components/community_admin.jsx]: The first lines are unnecessary, instead try this:

````
// this.props.params is set by react router
const { id } = this.props.params; // community id or name
const community = serverCache.getCommunity(id);
if (!community) {
    return <Wait/>;
}
if (community instanceof Error) {
    return <Err err={community}/>;
}
````

File: webui/src/components/community_admin.jsx [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+        var users = {};
+        if(roles.size > 0 ){ 
+            roles.forEach( role => { 
+                users[role.get('id')] = serverCache.getCommunityUsers(role.getIn(['id']));
+            })
+        }
+
+        var mergedUsers = {}; 
+        Object.keys(users).forEach(role => { 
+            if(typeof(users[role]) !== "undefined" && Object.keys(users[role]).length > 0 ){
```
Comment (emanueldima) [inline webui/src/components/community_admin.jsx]: I think this line can be replaced with just `if (users[role])`

File: webui/src/components/community_admin.jsx [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+    componentWillReceiveProps(nextProps){
+        this.setState({currentAssignedRole: nextProps.defRoleName});
+    },
+
+    onSubmit(e) {
+        e.preventDefault();
+        this.editUsersRoles();
+    },
+
+    assigneRole(roleid, userEmail){
```
Comment (emanueldima) [inline webui/src/components/community_admin.jsx]: typo

File: webui/src/components/community_admin.jsx [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+                notifications.success("The new role was added to the user");  
+            },
+            () => {
+                notifications.danger("An error occured while trying to add the new role to the user");
+            }
+        );
+    },
+
+    deleteRole(roleid, userid){
+        const current_user=serverCache.getUser()
```
Comment (emanueldima) [inline webui/src/components/community_admin.jsx]: please leave a space before and after `=`

Comment (nharraud) [inline webui/src/components/community_admin.jsx]: @emanueldima if we want to enforce some code style maybe we could use some tools enabling it? I was already planning to add pep8 code style verification at some point. In Invenio the style is checked when `./run-tests.sh` is ran.

File: webui/src/components/community_admin.jsx (hunk 13 above)
Comment (SarahBA) [inline webui/src/components/community_admin.jsx]: `if (users[role])` doesn't seem to be sufficient. When the `users` doesn’t have value yet, it’s either an empty object, or `undefined`.

File: webui/src/data/server.js (hunk 3 above)
Comment (SarahBA) [inline webui/src/data/server.js]: All the information needed for communities and roles can be retrieved by `this.props.params.id` and `serverCache.getCommunity(communityName)`! Don’t need to use `currentCommunityRoles` or `getCommunityRoles()` anymore.

**Pull request merged**